### ISSUE TITLE
profile bot runs

### ISSUE THREAD

**Issue by beckermr** [regro/cf-scripts#1066]
We should profile the run of the bot code to see where we are spending most of our time.

**Comment by viniciusdc** [comment]
Hi, what bout using `cProfile` to profile the information of the `run` and save it somewhere to check it locally with `KCacheGrind`. This way we can see besides the actual time the current workflow of the calls? It's also possible to create an interactive visualization at the web services using `Django and Chart.js` with that information.

**Comment by viniciusdc** [comment]
OK, besides just to check: when you say the `run of the bot code ` you mean the `run` function or all the `run_bot` process? (I should've said it before)

**Comment by beckermr** [comment]
all of the code that makes prs

**Comment by viniciusdc** [comment]
> all of the code that makes prs

I think that is just `update_PRs`, right ? The others at `run` and `make_graph` just load some informations.

**Comment by beckermr** [comment]
sounds right yes!

**Comment by CJ-Wright** [comment]
I think the main thing to profile is all the code in [auto_tick.py](https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/auto_tick.py) since this is where the PRs are made (and the migrations are actually performed).

**Comment by viniciusdc** [comment]
> I think the main thing to profile is all the code in [auto_tick.py](https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/auto_tick.py) since this is where the PRs are made (and the migrations are actually performed).

Yup, you are alright I thought about that but I was not sure if we would profile the migration too, thanks for the comment.

**Comment by CJ-Wright** [comment]
I would profile everything, we need a better view of what is slow so we can fix it.

**Comment by viniciusdc** [comment]
What do you think ? #1071

**Comment by beckermr** [comment]
i'd be great to see the profile data.

**Comment by viniciusdc** [comment]
> i'd be great to see the profile data.

Inside the client logs ?

**Comment by beckermr** [comment]
the profiler outputs data on what is taking time. i'd be great to have us all look at that data together.

**Comment by viniciusdc** [comment]
Yup, I agree with you, but I don't get your point, the code is supposed to output the data into a file and them use another tool to see it's information more graphically, but if it's needed to output the data at the terminal window it's also possible.  The idea to output the data, was cause this way we can monitor all the function calls and times in a graph, to compare the using memory and other informations

**Comment by beckermr** [comment]
yup. So can you run the profiler on the code and attach the data here?

**Comment by viniciusdc** [comment]
> yup. So can you run the profiler on the code and attach the data here?

it's possible, Matt is `conda_build` a package ?

**Comment by beckermr** [comment]
yes, you can install it via `conda install conda-build`

**Comment by viniciusdc** [comment]
> yes, you can install it via `conda install conda-build`

Oh, thanks I was not sure if `conda-build` and `conda_build` were the same

**Comment by viniciusdc** [comment]
Sorry for the delay, I am having some issues with relative imports. Do you know if this a common mistake I am doing ?

**Comment by beckermr** [comment]
I'd need a code sample. They only work within packages.

**Comment by viniciusdc** [comment]
An example could be `from .xonsh_utils import env` or all the relative imports for the `auto_tick`

**Comment by beckermr** [comment]
hmm that should work as long as you are in a package

**Comment by viniciusdc** [comment]
> hmm that should work as long as you are in a package

Yes,It's been happening a long time, but with minor scripts that I could just one by one remove the inconsistencies but for `auto_tick.py` there is a lot of subordinate files with relative imports (this is whay I started using CI)

**Comment by viniciusdc** [comment]
With a big help of CJ and Matt, we could finish the tests with the profiler and receive the first [outputs](https://github.com/regro/cf-graph-countyfair/tree/master/profiler) :tada:

**Comment by viniciusdc** [comment]
Now that it's working, we should try to visualize it, I recommend the SnakeViz because it's browser adaptation. Also, I will try to insert the `@profile` decorator for further usage.

**Comment by beckermr** [comment]
Great! Let's chat about the results tomorrow.

**Comment by viniciusdc** [comment]
I think we can set this done right ?

**Comment by beckermr** [comment]
Let's wait until we have profiled the PR json steps too.

**Comment by viniciusdc** [comment]
> Let's wait until we have profiled the PR json steps too.

If possible, give a check to #1110
